gtdynamics/utils/FootContactState.cpp

```diff
@@ -0,0 +1,111 @@
+/* ----------------------------------------------------------------------------
+ * GTDynamics Copyright 2020, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * See LICENSE for the license information
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file  FootContactState.cpp
+ * @brief Utility methods for generating FootContactState objects.
+ * @author: Disha Das, Frank Dellaert
+ */
+
+#include <gtdynamics/factors/ObjectiveFactors.h>
+#include <gtdynamics/utils/FootContactState.h>
+
+#include <iostream>
+
+namespace gtdynamics {
+
+using gtsam::Matrix;
+using gtsam::NonlinearFactorGraph;
+using gtsam::Point3;
+using std::string;
+
+FootContactState::FootContactState( const std::vector<PointOnLink> &points_on_links) {
+  for (auto &&point_on_link : points_on_links) {
+    contact_points_.push_back(point_on_link);
+  }
+}
+
+FootContactState::FootContactState(const std::vector<LinkSharedPtr> &links,
+                                   const gtsam::Point3 &contact_in_com) {
+  for (auto &&link : links) {
+    contact_points_.emplace_back(link, contact_in_com);
+  }
+}
+
+std::ostream &operator<<(std::ostream &os, const FootContactState &phase) {
+  os << "[";
+  for (auto &&cp : phase.contactPoints()) {
+    os << cp.link->name() << ": [" << cp.point.transpose() << "], ";
+  }
+  os << "]";
+  return os;
+}
+
+void FootContactState::print(const string &s) const {
+  std::cout << (s.empty() ? s : s + " ") << *this << std::endl;
+}
+
+// Searches a contact_link from ContactGoals object and returns the
+// corresponding goal_point
+gtsam::Point3 &pointGoal(ContactGoals *cp_goals,
+                         const PointOnLink contact_point) {
+  for (auto it = cp_goals->begin(); it != cp_goals->end(); ++it) {
+    if ((*it).point_on_link.link->name() == contact_point.link->name() &&
+        (*it).point_on_link.point == contact_point.point)
+      return (*it).goal_point;
+  }
+  throw std::runtime_error("Contact Point was not found.");
+}
+
+NonlinearFactorGraph FootContactState::contactPointObjectives(
```

Comment:
I think we might want to put this in KinematicsPhase instead?

Reply:
I am a bit confused by this. We have a pointGoalObjectives method under KinematicsPhase (which calls a similar method in KinematicsInterval). Both `FootContactState::contactPointObjectives` and `KInematicsPhase::pointGoalObjectives` push pointGoalFactors to the factor graph (differently). I am not sure what is the difference and when each of them is used.

Reply:
Something to chat about; maybe prepare one slide for meeting on refactored structure?